lib/bap/bap_self.ml

```diff
@@ -60,4 +60,73 @@ module Create() = struct
   let warning f = message Warning ~section:name f
   let error f = message Error ~section:name f
 
+  module Config = struct
+    let plugin_name = name
+    include Bap_config
+
+    let (/) = Filename.concat
+    let confdir = confdir / plugin_name
+    let conf_filename = confdir / "config"
+
+    exception Improper_format of string
+
+    let get_env_options () =
+      let prefix = "BAP_" ^ String.uppercase plugin_name ^ "_" in
+      let prefix_chop_key (k, v) =
+        match String.chop_prefix ~prefix k with
+        | Some k -> Some (k, v)
+        | None -> None
+      in
+      let plugin_filter_map str =
+        match String.split str ~on:'=' with
+        | [k; v] -> prefix_chop_key (k, v)
+        | _ -> None
+      in
+      Unix.environment () |>
+      Array.to_list |>
+      List.filter_map ~f:plugin_filter_map
+
+    let options () =
+      let string_splitter str =
+        match String.split str ~on:'=' with
+        | [k; v] -> (k, v)
+        | _ -> raise (Improper_format str)
+      in
+      let split_filter = List.map ~f:string_splitter in
+      let conf_file_options =
+        try
+          In_channel.with_file
+            conf_filename ~f:(fun ch -> In_channel.input_lines ch
+                                        |> split_filter)
+        with Sys_error _ -> []
+      in
+      get_env_options () |>
+      List.fold ~init:conf_file_options
+        ~f:(fun o (k, v) -> List.Assoc.add o k v)
+
+    let get name =
+      let name = String.uppercase name in
+      let search_for = options () |> List.Assoc.find in
+      search_for name
+
+    let set ~name ~data =
+      let name = String.uppercase name in
+      let old_conf = options () in
+      let new_conf = List.Assoc.add old_conf name data in
+      let conf_lines = List.map new_conf ~f:(fun (k, v) -> k ^ "=" ^ v) in
+      let write_lines () = Out_channel.write_lines
+          conf_filename conf_lines in
+      try
+        write_lines()
+      with Sys_error _ ->
+        let dir_permissions = 0o755 in
+        let makedir_and_write_lines () =
+          Unix.mkdir confdir dir_permissions; write_lines() in
+        try
+          makedir_and_write_lines()
```

Comment:
I don't see a big win in creating a one line function, and calling them inside the try/catch block, why not just put them body in it. Moreover, I don't think that using exceptions here for control flow is adding much to readability, given that you're making an assumption that a failed write to a file implies the absence of a parent directory. I think the following is more straightforward:
```ocaml
if not (Sys.file_exists conf_dir) 
then Unix.mkdir conf_dir conf_perm;
if not (Sys.is_directory conf_dir)
then invalid_argf "%s is not a directory";
Out_channel.write_lines filename config
```

This example also demonstrates a slightly different naming convention, that is used in OCaml. We use shorter names, as the scope of the name is usually much more limited, and referential transparency, strong typing and immutability reduces a possible impact of name clashes.